### PR TITLE
[dagster-aws] add Pipes cloudwatch message reader

### PR DESCRIPTION
**Pull request by danielgafni**
## Summary & Motivation

Adds PipesCloudWatchMessageReader which can be used with different AWS services. 

It's a solid default MessageReader since various AWS services already emit CloudWatch logs by default. 

Right now it's reading the full CloudWatch log stream provided by the user and routing it to Dagster's stdout. That's the minimal functionality needed by Glue Pipes. 

In the future, we can think about changing this behavior, mainly: 
- optionally, filtering non-dagster logs in the CloudWatch query
- optionally, route logs to stderr 

It's work mentioning that Glue routes both `stdout` and `stderr` to `/output` log group. Thus, we only need to read a single stream (with the job run id) from this group in order to receive all necessary information.  

## How I Tested These Changes
- [x]: add moto tests with emulating CloudWatch logs

### CODE REVIEW

**Comment by danielgafni** [comment]
* **#23354** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23354?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#23353** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23353?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @danielgafni and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/libraries/dagster-aws/dagster_aws/pipes.py [hunk 1]
```diff
@@ -157,10 +169,22 @@ def no_messages_debug_text(self) -> str:
...
+
 @experimental
 class PipesCloudWatchMessageReader(PipesMessageReader):
     """Message reader that consumes AWS CloudWatch logs to read pipes messages."""
 
+    def __init__(self, client: Optional[boto3.client] = None):
+        """Args:
+        client (boto3.client): boto3 CloudWatch client.
+        """
+        self.client = client or boto3.client("logs")
```
Comment (danielgafni) [inline python_modules/libraries/dagster-aws/dagster_aws/pipes.py:186]: I decided to automatically create default aws clients for the pipes components. 
AWS clients are usually configured via environment variables and local configs, so I think that's a sensible default. 

Is this ok to do?

Comment (schrockn) [reply python_modules/libraries/dagster-aws/dagster_aws/pipes.py:186]: Seems quite sensible! 👍🏻

File: docs/content/concepts/dagster-pipes/aws-glue.mdx [hunk 2]
```diff
@@ -46,24 +46,26 @@ Call `open_dagster_pipes` in the Glue job script to create a context that can be
...
 client = boto3.client("s3")
 context_loader = PipesS3ContextLoader(client)
-message_writer = PipesS3MessageWriter(client)
 params_loader = PipesCliArgsParamsLoader()
 
 
 def main():
     with open_dagster_pipes(
         context_loader=context_loader,
-        message_writer=message_writer,
```
Comment (schrockn) [inline docs/content/concepts/dagster-pipes/aws-glue.mdx:63]: So this is this our new recommendation? Writing messages to stdout and then doing it re: cloudwatch? What are the cost implications here?

**Review by schrockn — CHANGES_REQUESTED** [review]
This looks great overall. Needs more tests and a more flesh-out test plan.

Can you also provide a more thorough "summary and motivation" section about why this and the tradeoffs involved. If memory serves there are a real cost tradeoffs here and it would be good to understand that and document them.

File: docs/content/concepts/dagster-pipes/aws-glue.mdx (hunk 2 above)
Comment (danielgafni) [reply docs/content/concepts/dagster-pipes/aws-glue.mdx:63]: Well, AWS GLue jobs emit lots of Glue logs (via stdout/err) by default. 

We will be adding a small amount of dagster logs to these streams. I don't think it matters at all.

**Comment by danielgafni** [comment]
Side note: I discovered that `moto` + S3 tests fail if no credentials are configured. This is surprising a for a mocking software. Took me a while to figure this out on a new clean VM... 

Found this issue here https://github.com/getmoto/moto/issues/1941) describing similar situations. 

Apparently, it's necessary to have at least some credentials configured. 

Should we provide testing credentails for all our pytest tests automatically?

**Comment by danielgafni** [comment]
@schrockn hey, please take a look at `moto` tests. 

I had to manually create CloudWatch logs (since `moto` doesn't actually run the Glue job and thus doesnt create any logs) in the fake Glue client. 

I'm planning to move examples/docs changes into a separate PR tomorrow.

File: examples/docs_snippets/docs_snippets/guides/dagster/dagster_pipes/glue/dagster_code.py [hunk 3]
```diff
@@ -39,9 +39,6 @@ def glue_pipes_asset(
                 client=boto3.client("s3"),
```
Comment (schrockn) [inline examples/docs_snippets/docs_snippets/guides/dagster/dagster_pipes/glue/dagster_code.py:39]: <i class='graphite__hidden'>[Re: line -38]<br/><br/></i>

This predates this but does the `PipesGlueContextInjector` really provide enough value for its own class beyond `PipesS3ContextInjector`?<span class='graphite__hidden'><br/><br/>See this comment inline on <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23353?utm_source=unchanged-line-comment">Graphite</a>.</span>

File: python_modules/libraries/dagster-aws/dagster_aws/pipes.py [hunk 4]
```diff
@@ -174,13 +198,53 @@ def read_messages(
...
+        handler = check.not_none(
+            self._handler, "Can only consume logs within context manager scope."
+        )
+
+        for events_batch in self._get_all_cloudwatch_events(
+            log_group=log_group, log_stream=log_stream, start_time=start_time, end_time=end_time
+        ):
+            for event in events_batch:
+                for log_line in event["message"].splitlines():
+                    extract_message_or_forward_to_stdout(handler, log_line)
```
Comment (schrockn) [inline python_modules/libraries/dagster-aws/dagster_aws/pipes.py:216]: Ok so this means we will ingest all of stdout from cloudwatch. What about stderr?

**Review by schrockn — CHANGES_REQUESTED** [review]
I have some minor inline comments.

My big question here is what will appear in Dagster UI? We are now getting all the cloudwatch logs. Are they all going to show up in stdout? Where does stderr emitted from a glue job end up?

Paint me a picture of the end-to-end logging experience in Dagster in this case.

File: examples/docs_snippets/docs_snippets/guides/dagster/dagster_pipes/glue/dagster_code.py (hunk 3 above)
Comment (danielgafni) [reply examples/docs_snippets/docs_snippets/guides/dagster/dagster_pipes/glue/dagster_code.py:39]: Same as with the other dummy class, the only reason I've added it was to repeat the same pattern started with Lambda Pipes. 

I was (and still am) unsure if it was necessary. We decided to go for it with @cmpadden and @smackesey in the original Glue Pipes PR. 

I think we can remove these dummy classes. They serve some "discovery" (?) purpose for a beginner Pipes user trying to connect appropriate pipes pieces together, but I don't think this role is really important. 

I'm up to removing them.

File: python_modules/libraries/dagster-aws/dagster_aws/pipes.py (hunk 4 above)
Comment (danielgafni) [reply python_modules/libraries/dagster-aws/dagster_aws/pipes.py:216]: Funny enough, that's not how AWS Glue handles stdout/err. 

I haven't completely figured it out yet, but it seems like it actually routes stderr to `/out` LogGroup. Here is an example with some Dagster logs written by the default MessageWriter - that's stderr right? 

image

Meanwhile, `/error` LogGroup contains multiple streams: spark worker logs and the mains spark process logs. That's how it looks like: 

image

No trace of Python/Dagster stderr.

**Comment by danielgafni** [comment]
@schrockn As I suspected, AWS Glue routes both `stdout` and `stderr` to `/output` CloudWatch log group. The `/error` sterams in CloudWatch only contain Spark logs. 

The following code: 

```python
with open_dagster_pipes(
        params_loader=PipesCliArgsParamsLoader(),
        context_loader=PipesS3ContextLoader(client=boto3.client("s3"))
    ) as pipes:
    pipes.log.info("Hello from external process!")
    pipes.report_asset_materialization(
            metadata={
                "some_metric": {"raw_value": 0, "type": "int"}
            },
            data_version="alpha",
        )
    
    print("hello from stdout")
    print("hello from stderr", file=sys.stderr)
```

produces this: 

image

image

The line `"hello from stderr"` which is explicitly printed to `stderr` still ends up in the `/output` LogStream. 

I don' think we can (or should) do anything about it. This is just how Glue is.

**Comment by schrockn** [comment]
> @schrockn As I suspected, AWS Glue routes both `stdout` and `stderr` to `/output` CloudWatch log group. The `/error` sterams in CloudWatch only contain Spark logs. 
> 
> The following code: 
> 
> ```python
> with open_dagster_pipes(
>         params_loader=PipesCliArgsParamsLoader(),
>         context_loader=PipesS3ContextLoader(client=boto3.client("s3"))
>     ) as pipes:
>     pipes.log.info("Hello from external process!")
>     pipes.report_asset_materialization(
>             metadata={
>                 "some_metric": {"raw_value": 0, "type": "int"}
>             },
>             data_version="alpha",
>         )
>     
>     print("hello from stdout")
>     print("hello from stderr", file=sys.stderr)
> ```
> 
> produces this: 
> 
> image
> 
> image
> 
> The line `"hello from stderr"` which is explicitly printed to `stderr` still ends up in the `/output` LogStream. 
> 
> I don' think we can (or should) do anything about it. This is just how Glue is. 

Yup 100% agree. Can you document this behavior in both the PR summary and as a docblock?

**Comment by schrockn** [comment]
In terms of eliminating the glue-specific wrapper reader/writer classes, can you do a seperate PR stacked on this with colton included so we can discuss it separately?

**Review by schrockn — CHANGES_REQUESTED** [review]
request for docs

**Comment by danielgafni** [comment]
I believe this PR should be ready

File: python_modules/libraries/dagster-aws/dagster_aws_tests/pipes_tests/test_pipes.py [hunk 5]
```diff
@@ -293,12 +295,18 @@ def script_fn():
...
         )
 
-        client = boto3.client("s3", region_name="us-east-1", endpoint_url="http://localhost:5193")
-        context_loader = PipesS3ContextLoader(client=client)
-        message_writer = PipesS3MessageWriter(client, interval=0.001)
+        s3_client = boto3.client(
+            "s3", region_name="us-east-1", endpoint_url="http://localhost:5193"
+        )
+
+        messages_backend = os.environ["PIPES_MESSAGES_BACKEND"]
```
Comment (schrockn) [inline python_modules/libraries/dagster-aws/dagster_aws_tests/pipes_tests/test_pipes.py:302]: is this an existing env var? only in test?

File: python_modules/libraries/dagster-aws/dagster_aws/pipes.py [hunk 6]
```diff
@@ -399,11 +453,25 @@ def run(
...
+                    f"Glue job {job_name} run {run_id} failed:\n{response['JobRun']['ErrorMessage']}"
+                )
+            else:
+                context.log.info(f"Glue job {job_name} run {run_id} completed successfully")
+
+            if isinstance(self._message_reader, PipesCloudWatchMessageReader):
+                # TODO: receive messages from a background thread in real-time
+                self._message_reader.consume_cloudwatch_logs(
+                    f"{log_group}/output", run_id, start_time=int(start_timestamp)
+                )
```
Comment (schrockn) [inline python_modules/libraries/dagster-aws/dagster_aws/pipes.py:473]: can you expand on this TODO so a person without context understands what this is about?

**Review by schrockn — APPROVED** [review]

File: python_modules/libraries/dagster-aws/dagster_aws_tests/pipes_tests/test_pipes.py (hunk 5 above)
Comment (danielgafni) [reply python_modules/libraries/dagster-aws/dagster_aws_tests/pipes_tests/test_pipes.py:302]: This is only for testing (to parametrize over message passing options). I will rename it to `TESTING_PIPES_MESSAGES_BACKEND` to remove confusion.

File: python_modules/libraries/dagster-aws/dagster_aws/pipes.py (hunk 6 above)
Comment (danielgafni) [reply python_modules/libraries/dagster-aws/dagster_aws/pipes.py:473]: done

**Pull request merged**